pghoard/transfer.py

```diff
@@ -309,25 +310,27 @@ def handle_download(self, site, key, file_to_transfer):
             self.metrics.unexpected_exception(ex, where="handle_download")
             return CallbackEvent(success=False, exception=ex, opaque=file_to_transfer.opaque)
 
-    def handle_upload(self, site, key, file_to_transfer):
+    def handle_upload(self, site, key, file_to_transfer: UploadEvent):
         payload = {"file_size": file_to_transfer.file_size}
         try:
             storage = self.get_object_storage(site)
             unlink_local = file_to_transfer.remove_after_upload
             self.log.info("Uploading file to object store: src=%r dst=%r", file_to_transfer.source_data, key)
             if not isinstance(file_to_transfer.source_data, BytesIO):
-                f = open(file_to_transfer.source_data, "rb")
+                f = open(file_to_transfer.source_data, "rb")  # type: ignore
             else:
                 f = file_to_transfer.source_data
             with f:
                 metadata = file_to_transfer.metadata.copy()
                 if file_to_transfer.file_size:
-                    metadata["Content-Length"] = file_to_transfer.file_size
-                storage.store_file_object(key, f, metadata=metadata)
+                    metadata["Content-Length"] = file_to_transfer.file_size  # type: ignore
+                storage.store_file_object(
+                    key, f, metadata=metadata, upload_progress_fn=file_to_transfer.incremental_progress_callback
+                )
             if unlink_local:
                 try:
                     self.log.info("Deleting file: %r since it has been uploaded", file_to_transfer.source_data)
-                    os.unlink(file_to_transfer.source_data)
+                    os.unlink(file_to_transfer.source_data)  # type: ignore
```

Comment:
I'd rather have an if `isinstance(file_to_transfer.source_data, Path)` than a ignored typed check here.

Adding something like that in the `UploadTransfer` definition would help catch the error earlier too:
```python
    def __post_init__(self) -> None:
        if self.remove_after_upload and not isinstance(self.source_data, Path):
            raise ValueError("remove_after_upload can only be True if the source is a Path")
```